Improve series, sample, metadata and exemplars validation errors (#1907)
grafana/mimir: operations/mimir-mixin/docs/playbooks.md

  1. The alert fired because of a bug in Mimir: fix it.
  1. The alert fired because of a bug or edge case in the continuous test tool, causing a false positive: fix it.

## Codified errors

Mimir has some codified error IDs that you might see in HTTP responses or logs.

> These error IDs allow you to quickly look up more details in the documentation below.

Reviewer: Replace directional language (below, above, left, or right) with a term that applies to any device on which one is reading the text.

### err-mimir-missing-metric-name

> This non-critical error occurs when Mimir receives a write request containing a series without metric name.
> The metric name is required for each series.

Reviewer: ```suggestion
This non-critical error occurs when Mimir receives a write request that contains a series that does not have a metric name.
You must provide a metric name for each series.
```
A assume a human if providing the metric name rather than a machine.

Author: Metrics are machine generated. Specs require that each metric has a name.

Reviewer: By including this message that a machine generates, "The metric name is required for each series.", how does the operator supply one if it does not exist?

Author: If metrics name is missing it's typically a bug of the sender client, so should be reported. There could be edge cases where the operator is a developer, writing their own tool to write metrics, and in that case they're in control of that bug too (but that's an edge case).

Reviewer: I added context that you can decide to keep or not keep.

> > **Note**: Invalid series are skipped during the ingestion, while other valid series within the same request are ingested.

Reviewer: ```suggestion
> **Note**: Invalid series are skipped during the ingestion, and other valid series within the same request are ingested.
```

### err-mimir-metric-name-invalid

This non-critical error occurs when Mimir receives a write request containing a series with an invalid metric name.

> A metric name can only contain a well known set of characters, as defined by [Prometheus specifications](https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels).

Reviewer: Also, make the link text as WYSIWYG as possible.

Specifically, it may only contain ASCII letters, digits, underscores and colons.

> > **Note**: Invalid series are skipped during the ingestion, while other valid series within the same request are ingested.

Reviewer: ```suggestion
> **Note**: Invalid series are skipped during the ingestion, and valid series within the same request are ingested.
```

### err-mimir-max-label-names-per-series

> This non-critical error occurs when Mimir receives a write request containing a series with a number of labels exceeding the configured limit.

Reviewer: ```suggestion
This non-critical error occurs when Mimir receives a write request that contains a series with a number of labels that exceed the configured limit.
```

Reviewer 2: Should this say _"...with a number of label **names** that exceed..."_ for consistency with the error and the option?

> The limit is used to protect the system stability from potential abuses or mistakes, and it can be configured on a per-tenant basis using `-validation.max-label-names-per-series`.
> 

Reviewer: ```suggestion
The limit protects the system’s stability from potential abuse or mistakes, and you can configure the limit on a per-tenant basis by using the `-validation.max-label-names-per-series` option.

```

Reviewer: Assume one system instead of more than one.

> > **Note**: Invalid series are skipped during the ingestion, while other valid series within the same request are ingested.

Reviewer: ```suggestion
> **Note**: Invalid series are skipped during the ingestion, and valid series within the same request are ingested.
```

> ### err-mimir-label-invalid

Reviewer 2: I'd make this `err-mimir-label-name-invalid`, to be consistent with the error, which fails when label name is invalid.

Author: Error IDs were kept to match existing `reason` label values used in metrics, to have a 1:1 matching.

This non-critical error occurs when Mimir receives a write request containing a series with an invalid label name.
A label name can only contain a well known set of characters, as defined by [Prometheus specifications](https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels).
Specifically, it may only contain ASCII letters, digits, and underscores.

> 
> > **Note**: Invalid series are skipped during the ingestion, while other valid series within the same request are ingested.
> 
> ### err-mimir-label-name-too-long
> 
> This non-critical error occurs when Mimir receives a write request containing a series with a label name whose length exceeds the configured limit.
> The limit is used to protect the system stability from potential abuses or mistakes, and it can be configured on a per-tenant basis using `-validation.max-length-label-name`.
> 
> > **Note**: Invalid series are skipped during the ingestion, while other valid series within the same request are ingested.
> 
> ### err-mimir-label-value-too-long
> 
> This non-critical error occurs when Mimir receives a write request containing a series with a label value whose length exceeds the configured limit.
> The limit is used to protect the system stability from potential abuses or mistakes, and it can be configured on a per-tenant basis using `-validation.max-length-label-value`.
> 
> > **Note**: Invalid series are skipped during the ingestion, while other valid series within the same request are ingested.
> 
> ### err-mimir-duplicate-label-names
> 
> This non-critical error occurs when Mimir receives a write request containing a series with the same label name repeated two or more times.
> A series containing repeated label names is invalid and so it gets skipped during the ingestion.
> 
> > **Note**: Invalid series are skipped during the ingestion, while other valid series within the same request are ingested.
> 
> ### err-mimir-labels-not-sorted
> 
> This error occurs when Mimir receives a write request containing a series whose label names are not sorted alphabetically.
> However, Mimir internally sort labels for received series, so this error shouldn't occur in pratice.
> If you experience this error, please open an issue in the Mimir repository.
> 
> > **Note**: Invalid series are skipped during the ingestion, while other valid series within the same request are ingested.
> 
> ### err-mimir-too-far-in-future
> 
> This non-critical error occurs when Mimir receives a write request containing a sample whose timestamp is in the future compared to the current "real world" time.
> Mimir tolerates timestamps slightly in the future (e.g. due to clock screws), while rejects timestamps more then `-validation.create-grace-period` in the future.
> You can fine tune the toleration on a per-tenant basis using `-validation.max-length-label-value`.
> 
> > **Note**: Series with invalid samples are skipped during the ingestion, while other valid series within the same request are ingested.
> 
> ### err-mimir-exemplar-labels-missing
> 
> This non-critical error occurs when Mimir receives a write request containing an exemplar without any label identifying the related metric.
> An exemplar must have at least a valid label pairs, otherwise it can't be associated to any metric.
> 
> > **Note**: Invalid exemplars skipped during the ingestion, while other valid exemplars within the same request are ingested.
> 
> ### err-mimir-exemplar-labels-too-long
> 
> This non-critical error occurs when Mimir receives a write request containing an exemplar whose combined labels set size exceeds the limit.
> The limit is used to protect the system stability from potential abuses or mistakes, and it can't be configured.
> 
> > **Note**: Invalid exemplars skipped during the ingestion, while other valid exemplars within the same request are ingested.
> 
> ### err-mimir-exemplar-timestamp-invalid
> 
> This non-critical error occurs when Mimir receives a write request containing an exemplar without timestamp.
> An exemplar must have a valid timestamp, otherwise it can't be correlated to any point in time.
> 
> > **Note**: Invalid exemplars skipped during the ingestion, while other valid exemplars within the same request are ingested.

Reviewer: ```suggestion
### err-mimir-label-invalid

This non-critical error occurs when Mimir receives a write request that contains a series with an invalid label name.
A label name name can only contain characters as defined by Prometheus’ [Metric names and labels](https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels).

> **Note**: Invalid series are skipped during the ingestion, and valid series within the same request are ingested.

### err-mimir-label-name-too-long

This non-critical error occurs when Mimir receives a write request that contains a series with a label name whose length exceeds the configured limit.
The limit protects the system’s stability from potential abuse or mistakes, and you can configure the limit on a per-tenant basis by using the `-validation.max-length-label-name` option.

> **Note**: Invalid series are skipped during the ingestion, and valid series within the same request are ingested.

### err-mimir-label-value-too-long

This non-critical error occurs when Mimir receives a write request that contains a series with a label value whose length exceeds the configured limit.
The limit protects the system’s stability from potential abuse or mistakes, and you can configure the limit on a per-tenant basis by using the `-validation.max-length-label-value` option.

> **Note**: Invalid series are skipped during the ingestion, and valid series within the same request are ingested.

### err-mimir-duplicate-label-names

This non-critical error occurs when Mimir receives a write request that contains a series with the same label name two or more times.
A series that contains a duplicated label name is invalid and gets skipped during the ingestion.

> **Note**: Invalid series are skipped during the ingestion, and valid series within the same request are ingested.

### err-mimir-labels-not-sorted

This error occurs when Mimir receives a write request that contains a series whose label names are not sorted alphabetically.
However, Mimir internally sorts labels for series that it receives, so this error should not occur in practice.
If you experience this error, [open an issue in the Mimir repository](https://github.com/grafana/mimir/issues).

> **Note**: Invalid series are skipped during the ingestion, and valid series within the same request are ingested.

### err-mimir-too-far-in-future

This non-critical error occurs when Mimir receives a write request that contains a sample whose timestamp is in the future compared to the current "real world" time.
Mimir accepts timestamps that are slightly in the future, due to skewed clocks for example. It rejects timestamps that are too far in the future, based on the definition that you can set via the `-validation.create-grace-period` option.
On a per-tenant basis, you can fine tune the tolerance by configuring the `-validation.max-length-label-value` option.

> **Note**: Series with invalid samples are skipped during the ingestion, and series within the same request are ingested.

### err-mimir-exemplar-labels-missing

This non-critical error occurs when Mimir receives a write request that contains an exemplar without a label that identifies the related metric.
An exemplar must have at least one valid label pair, otherwise it cannot be associated with any metric.

> **Note**: Invalid exemplars are skipped during the ingestion, and valid exemplars within the same request are ingested.

### err-mimir-exemplar-labels-too-long

This non-critical error occurs when Mimir receives a write request that contains an exemplar where the combined set size of its labels exceeds the limit.
The limit is used to protect the system’s stability from potential abuse or mistakes, and it cannot be configured.

> **Note**: Invalid exemplars are skipped during the ingestion, and valid exemplars within the same request are ingested.

### err-mimir-exemplar-timestamp-invalid

This non-critical error occurs when Mimir receives a write request that contains an exemplar without a timestamp.
An exemplar must have a valid timestamp, otherwise it cannot be correlated to any point in time.

> **Note**: Invalid exemplars are skipped during the ingestion, and valid exemplars within the same request are ingested.
```

Reviewer: Well, that makes reviewing difficult. I made similar changes to those previous to these lines. I will make a comment about the ones that deviate from repetition to make the review easier.
